instrumentation/excon/lib/opentelemetry/instrumentation/excon/patches/socket.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Excon
+      module Patches
+        # Module to prepend to an Excon Socket for instrumentation
+        module Socket
+          private
+
+          def connect
+            return super if untraced?
+
+            if @data[:proxy]
+              conn_address = @data[:proxy][:hostname]
+              conn_port = @data[:proxy][:port]
+            else
+              conn_address = @data[:hostname]
+              conn_port = @port
+            end
+
+            attributes = { OpenTelemetry::SemanticConventions::Trace::NET_PEER_NAME => conn_address, OpenTelemetry::SemanticConventions::Trace::NET_PEER_PORT => conn_port }.merge!(OpenTelemetry::Common::HTTP::ClientContext.attributes)
+
+            if is_a?(::Excon::SSLSocket) && @data[:proxy]
```

Comment:
Are these reusable by the same socket object instance? 

Is it possible to memoize this decision in some way?

Reply:
A socket is only ever connected once. We could potentially make this decision here instead: https://github.com/excon/excon/blob/v0.104.0/lib/excon/connection.rb#L470-L479

However, that would require a patch on the Connection as well as the Socket class. Also, memoizing it would be difficult because the inputs to the decision can change at any point so we would need to handle those updates.